File: ProyectoCivitasRuby/lib/modelo/mazo_sorpresas.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class MazoSorpresas
    
    #Constructor
    def initialize(d = false)
      init
      @debug = d
      if(d)
        Diario.instance.ocurre_evento("MazoSorpresas en modo debug")
      end
    end

    
  
    private
    def init
      #Atributos de referencia
      @sorpresas = Array.new   
      @cartasespeciales = Array.new
    
      #Atributos de instancia (algunos)
      @barajada = false
      @usadas = 0
      @ultimasorpresa = nil
    end
    
    
    
    public
    
    attr_reader :ultimasorpresa
    
    
    def al_mazo(s)
      if(@barajada==false)
        @sorpresas.push(s)
      end
    end
    
    
    def siguiente()
      if(@barajada==false || @usadas == @sorpresas.size())
        if(@debug==false)
          @sorpresas.shuffle!()
        end
        @usadas = 0
        @barajada = true
      end
      @usadas=@usadas+1
      @ultimasorpresa = @sorpresas[0]
      @sorpresas.delete_at(0)
      @sorpresas.push(@ultimasorpresa)
      
      return @ultimasorpresa 
    end
  
    
    def inhabilitar_carta_especial(s)
      if(@sorpresas.index!=nil)  
        @sorpresas.delete(s)
        @cartasespeciales.push(s)
        Diario.instance.ocurre_evento("Carta Especial inhabilitada")
      end
    end
      
    
    def habilitar_carta_especial(s)
      if(@cartasespeciales.index(s))
        @cartasespeciales.delete(s)
        @sorpresas.push(s)
        Diario.instance.ocurre_evento("Carta Especial habilitada")
      end
    end
    
  end
end

File: ProyectoCivitasJava/src/civitas/MazoSorpresas.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3 - Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package civitas;
import java.util.ArrayList;
import java.util.Collections;

public class MazoSorpresas {
    
    //Atributos de instancia
    private boolean barajada;
    private int usadas; 
    private boolean debug;
    private Sorpresa ultimaSorpresa;
    
    //Atributos de referencia
    public ArrayList <Sorpresa> sorpresas;
    private ArrayList <Sorpresa> cartasEspeciales;

    private void init(){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(boolean d){
        debug = d;
        init();
        if (debug){
            Diario.getInstance().ocurreEvento("MazoSorpresas en modo debug");
        }
    }
    
    MazoSorpresas(){
        init();
        debug = false;
    }
    
    void alMazo(Sorpresa s){
        if(!barajada){
            sorpresas.add(s);
        }
    }
    
    Sorpresa siguiente(){
        if(!barajada || usadas==sorpresas.size()){
            if(!debug){
                Collections.shuffle(sorpresas);
            }
            usadas = 0;
            barajada = true;
        }
        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);
     
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if(sorpresas.contains(sorpresa)){
            sorpresas.remove(sorpresa);
            cartasEspeciales.add(sorpresa);       
            Diario.getInstance().ocurreEvento("Carta Especial inhabilitada.");
        }
        

    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        if(cartasEspeciales.contains(sorpresa)){
            cartasEspeciales.remove(sorpresa);
            sorpresas.add(sorpresa);
            Diario.getInstance().ocurreEvento("Carta Especial habilitada");
        }        
    }
    
    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }
}
